### DISCUSSION TITLE
v0.6.15, struggling to set the "download" <a> attribute dynamically

### DISCUSSION THREAD

**Discussion by ViciousBadger** [leptos-rs/leptos#3109] [Q&A] [answered]
Hello, hope this is the right place to ask such a question.

I'm creating a basic wrapper component for an anchor element that is sometimes a download link, sometimes a link to an external site..

From what I can tell, if you write an attribute like `attr:download=(some bool)`, the attribute should show when the bool is true (as a simple value-less attribute). The source code confirms this: 

```
impl Attribute {
    [...]
    pub fn as_value_string [...] {
            [...]
            Attribute::Bool(include) => {
                (if *include { attr_name } else { "" })
            }
        }
    }
```

.. but when I try to do this in a view macro (please excuse the pseduocode):

```
    view! {
      <a
        role="button"
        href=target
        attr:download=true
      >
        {text}
      </a>
    }
```

The resulting HTML is `<a role="button" href=[...] download="download">[...]</a>`, which causes incorrect behaviour as the attribute value will be used as the downloaded file name, not what I want obviously.. I would like it to render as `<a role="button" href=[...] download>[...]</a>` am I doing something wrong or is this some kind of flaw in the view macro..?

**Comment by gbj** [comment, marked as answer]
> From what I can tell, if you write an attribute like `attr:download=(some bool)`, the attribute should show when the bool is true (as a simple value-less attribute). The source code confirms this:

Unfortunately I think you might be misreading the source code: the snippet you posted sets the attribute *value* for a boolean attribute to `if *include { attr_name }`, and later checks to see if the value is empty. So, it renders `download="download"` here.

Note that this is perfectly valid HTML for a [Boolean attribute](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML), for which `foo`, `foo=""`, and `foo="foo"` are all supposed to behave the same.

I have no idea why they designed [`download`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download) to work a bit differently than that. Experimenting a little shows me that 
* `<a href="/foo" download>` downloads a file named `foo.html`
* `<a href="/foo" download="">` downloads a file named `foo.html`
* `<a href="/foo" download="download">` downloads a file named `download.html`

Fair enough.

This suggests a (hopefully easy?) solution to me, though: rather than setting `download` to `true`, set it to `""`.

If it needs to be conditional, you can use `Option<_>` such that `false` (don't download) is `None` (in which case the attribute shouldn't appear at all), and `true` (do download) is `Some("")` (should appear as `download=""` which should work).

0.7 (git main) adopts the more minimal approach for boolean attributes of simply using the attribute name, for what it's worth.

**Comment by ViciousBadger** [comment]
Setting it to an empty string is a brilliant solution, thanks a ton. The behavior of "download" is definitely a bit weird.. not anywhere near the strangest quirks of HTML, but still.

Glad to hear about the 0.7 change, it does seem more sensible overall. With some luck it will _not_ introduce new weird hyper-specific issues !
